File: api-dot-net/Services/CodeExecution/Worker/CodeExecutionWorker.cs

namespace CjsApi.Services.CodeExecution.Worker
{

    using System.Threading.Channels;
    using CjsApi.Services.CodeExecution.Dto;
    using CjsApi.Services.CodeExecution.Factory;

    public class CodeExecutionWorker : BackgroundService
    {
        private readonly Channel<(ExecutionJob Job, CodeExecutionRequest Request)> _queue =
            Channel.CreateUnbounded<(ExecutionJob, CodeExecutionRequest)>();

        private readonly ICodeExecutorFactory _executorFactory;

        public CodeExecutionWorker(ICodeExecutorFactory executorFactory)
        {
            _executorFactory = executorFactory;
        }

        public void Enqueue(ExecutionJob job, CodeExecutionRequest request)
        {
            _queue.Writer.TryWrite((job, request));
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            await foreach (var (job, request) in _queue.Reader.ReadAllAsync(stoppingToken))
            {
                try
                {
                    job.Status = ExecutionStatus.Running;

                    var executor = _executorFactory.GetExecutor(request.Language);
                    job.Result = await executor.ExecuteAsync(request, stoppingToken);

                    job.Status = ExecutionStatus.Completed;
                }
                catch (Exception ex)
                {
                    job.Status = ExecutionStatus.Failed;
                    job.Error = ex.Message;
                }
            }
        }
    }

}

File: java-api/cjs-service/src/main/java/com/cjs/cjs_service/service/codeExecutionSerivce/worker/CodeExecutionWorker.java
package com.cjs.cjs_service.service.codeExecutionSerivce.worker;

import com.cjs.cjs_service.dto.CodeExecutionResult;
import com.cjs.cjs_service.dto.ExecutionJob;
import com.cjs.cjs_service.dto.ExecutionStatus;
import com.cjs.cjs_service.service.SubmissionService;
import com.cjs.cjs_service.service.codeExecutionSerivce.CodeExecutionRequest;
import com.cjs.cjs_service.service.codeExecutionSerivce.CodeExecutorFactory;

import jakarta.annotation.PostConstruct;

import org.springframework.boot.context.event.ApplicationReadyEvent;
import org.springframework.context.event.EventListener;
import org.springframework.scheduling.annotation.Async;
import org.springframework.stereotype.Component;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;

@Component
public class CodeExecutionWorker {

    private final BlockingQueue<JobRequest> queue = new LinkedBlockingQueue<>();

    private final CodeExecutorFactory executorFactory;
    private final SubmissionService submissionService;

    public CodeExecutionWorker(
            CodeExecutorFactory executorFactory,
            SubmissionService submissionService) {
        this.executorFactory = executorFactory;
        this.submissionService = submissionService;
    }

    public void enqueue(ExecutionJob job, CodeExecutionRequest request) {
        queue.offer(new JobRequest(job, request));
    }

    @EventListener(ApplicationReadyEvent.class)
    public void startWorker() {
        Thread.ofVirtual().start(this::workerLoop);
    }

    private void workerLoop() {
        while (!Thread.currentThread().isInterrupted()) {
            try {
                JobRequest jr = queue.take();
                ExecutionJob job = jr.job();
                CodeExecutionRequest request = jr.request();

                job.setStatus(ExecutionStatus.RUNNING);

                var executor = executorFactory.getExecutor(request.getLanguage());
                CodeExecutionResult result = executor.execute(request);

                job.setResult(result);

                if (job.getSubmissionId() != -1) {
                    submissionService.updateResult(
                            job.getSubmissionId(),
                            result.getSubmissionStatus(),
                            0,
                            0);
                }

                job.setStatus(ExecutionStatus.COMPLETED);

            } catch (InterruptedException e) {
                Thread.currentThread().interrupt(); // graceful shutdown
            } catch (Exception e) {
                e.printStackTrace();
            }
        }
    }

    private record JobRequest(ExecutionJob job, CodeExecutionRequest request) {
    }
}
